src/null-safety/understanding-null-safety/index.md

```diff
@@ -0,0 +1,1410 @@
+that uses the index operator and checks for `null` to see if the key is absent
+is very common, around half of all uses based on our analysis. Breaking all of
+that code would have set the Dart ecosystem aflame.
+
+Instead, the runtime behavior is the same and thus the return type is obliged
+to be nullable. This means you generally cannot immediately use the result of
+a map lookup:
+
+```dart
+var map = {"key": "value"};
+print(map["key"].length); // Error.
+```
+
+This gives you a compile error on the attempt to call `.length` on a nullable
+string. In cases where you *know* the key is present you can teach the type
+checker by using `!`:
+
+```dart
+var map = {"key": "value"};
+print(map["key"]!.length); // OK.
+```
+
+We considered adding another method to Map that would do this for you: look up
+the key, throw if not found, or return a non-nullable value otherwise. But what
+to call it? No name would be shorter than the single-character `!`, and no
+method name would be clearer than seeing a `!` with its built-in semantics right
+there at the call site. So the idiomatic way to access a known-present element
+in a map is to use `[]!`. You get used to it.
+
+### No unnamed List constructor
+
+The unnamed constructor on `List` creates a new list with the given size but
+does not initialize any of the elements. This would poke a very large hole in
+the soundness guarantees if you created a list of a non-nullable type and then
+accessed an element.
+
+To avoid that, we have removed the constructor entirely. It is an error to call
+`List()` in null safe code, even with a nullable type. That sounds scary, but
+in practice most code creates lists using list literals, `List.filled()`,
+`List.generate()`, or as a result of transforming some other collection. For
+the edge case where you want to create an empty list of some type, we added a
+new `List.empty()` constructor.
+
+The pattern of creating a completely uninitialized list has always felt out of
+place in Dart, and now it is even more so. If you have code broken by this,
+you can always fix it by using one of the many other ways to produce a list.
+
+### Cannot set a larger length on non-nullable lists
+
+This is little known, but the `length` getter on `List` also has a corresponding
+*setter*. You can set the length to a shorter value to truncate the list. And
+you can also set it to a *longer* length to pad the list with uninitialized
+elements.
+
+If you were to do that with a list of a non-nullable type, you'd violate
+soundness when you later accessed those unwritten elements. To prevent that, the
+`length` setter will throw a runtime exception if (and only if) the list has a
+non-nullable element type *and* you set it to a *longer* length. It is still
+fine to truncate lists of all types, and you can grow lists of nullable types.
+
+There is an important consequence of this if you define your own list types that
```

Comment:
This feels a little oddly stated.  The core issue is that you have to implement `add` yourself, because the version in ListMixin has to use the length setter.  I think `insert` actually may be fixed to not use `add` now?  cc @natebosch was this fixed?

Reply:
No, we have to use `add` for null safe code. Part of the migration does need to be to implement `add` if you are using `ListMixin`.

Reply:
Reworded it some.